### PR TITLE
[ENH] Add Basic ARIMA model

### PR DESCRIPTION
**Pull request by alexbanwell1**
#### Reference Issues/PRs
Split out of #2828
Basis for #2838

#### What does this implement/fix? Explain your changes.
Adds basic ARIMA forecaster, which fits phi, theta (and optionally c) parameters, but takes in the model to use e.g. ARIMA(p,d,q) and an option to include a constant term.

I have restructured this for numba purposes. The parameters were found using a bespoke implementation of Nelder Mead, sensibly generalised for any loss function. However, doing this means it cannot be cached, which massively slows it down. I have moved Nelder Mead back into the forecasting module, and have introduced an integer switch dispatch function to facilitate caching. 

This currently means some duplication of code in the loss_function and in ARIMA. This is usually the price for writing numba.

1. fit(y) sets the parameters for AR and MA component stores the one ahead forecast in self.forecast_
2. predict(y) applies the model to assumed unseen data, setting the MA component to zero as is standard (but a bit wierd). It returns the one ahead forecast for y.
3. forecast(y) fits on y and returns the one ahead forecast, using the residuals on y
4. iterative_forecast(y, prediction_horizon=12) fits on y, sets the residuals as to zero after tail of y exceeded.

```python
from aeon.datasets import load_airplane
from aeon.forecasting import ARIMA
y = load_forecasting()
a = ARIMA(p=2, d=1, q=2)
a.fit(y)
p1 = a.forecast_
p2 = a.forecast(y)
preds = a.iterative_forecast(y, prediction_horizon=12)
```

I have also moved the utils files into forecasting, since they are experimental

### CODE REVIEW

File: aeon/forecasting/_arima.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+    ----------
+    data_ : list of float
+        Original training series values.
+    differenced_data_ : list of float
+        Differenced version of the training data used for stationarity.
+    residuals_ : list of float
+        Residual errors from the fitted model.
+    aic_ : float
+        Akaike Information Criterion for the selected model.
+    p_, d_, q_ : int
```
[the remark this answers is not in the document]
Comment (alexbanwell1) [reply aeon/forecasting/_arima.py:42]: They are, but the attributes are left there so that they can by set in fit by the auto version, and allow it to inherit the attributes and methods.

File: aeon/forecasting/_arima.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+            A time series on which to learn a forecaster to predict horizon ahead
+        exog : np.ndarray, default =None
+            Optional exogenous time series data assumed to be aligned with y
+
+        Returns
+        -------
+        self
+            Fitted ARIMAForecaster.
+        """
+        self.p_ = self.p
```
[the remark this answers is not in the document]
Comment (alexbanwell1) [reply aeon/forecasting/_arima.py:109]: For the Auto version - see above

File: aeon/forecasting/_arima.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+@njit(cache=True, fastmath=True)
+def _aic(residuals, num_params):
+    """Calculate the log-likelihood of a model."""
+    variance = np.mean(residuals**2)
+    liklihood = len(residuals) * (np.log(2 * np.pi) + np.log(variance) + 1)
+    return liklihood + 2 * num_params
+
+
+@njit(fastmath=True)
+def _arima_model_wrapper(params, data, model):
```
[the remark this answers is not in the document]
Comment (alexbanwell1) [reply aeon/forecasting/_arima.py:171]: Yes this allows it to use the same code for the seasonal model, unfortunately for numba to work, it has to be a defined function, rather than a lambda, or factory method, which is a bit annoying!

File: aeon/forecasting/_arima.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+        """
+        y = np.array(y, dtype=np.float64)
+        value = _calc_arima(
+            self.differenced_data_,
+            self.model_,
+            len(self.differenced_data_),
+            _extract_params(self.parameters_, self.model_),
+            self.residuals_,
+        )
+        history = self.data_[::-1]
```
[the remark this answers is not in the document]
Comment (alexbanwell1) [reply aeon/forecasting/_arima.py:155]: Good point, currently it's only doing 1 ahead, will need to adjust this!

File: aeon/utils/forecasting/_hypo_tests.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
+import numpy as np
+
+
+def kpss_test(y, regression="c", lags=None):  # Test if time series is stationary
```
[the remark this answers is not in the document]
Comment (alexbanwell1) [reply aeon/utils/forecasting/_hypo_tests.py:4]: I guess so, but that value is not calculated - that's too difficult really for what we're trying to do, so it would have to be a look up table, which seems fairly pointless, for what we're trying to achieve here!

File: aeon/utils/forecasting/_hypo_tests.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    sigma_squared = gamma_0 + 2 * np.sum([w * g for w, g in zip(weights, gamma)])
+
+    # Step 4: Calculate the KPSS statistic
+    kpss_stat = np.sum(S_t**2) / (n**2 * sigma_squared)
+
+    if regression == "ct":
+        # p. 162 Kwiatkowski et al. (1992): y_t = beta * t + r_t + e_t,
+        # where beta is the trend, r_t a random walk and e_t a stationary
+        # error term.
+        crit = 0.146
```
[the remark this answers is not in the document]
Comment (alexbanwell1) [reply aeon/utils/forecasting/_hypo_tests.py:57]: 5%, but will add a comment in the code to make that clear

File: aeon/forecasting/_arima.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+    on information criteria, such as AIC.
+
+    Parameters
+    ----------
+    horizon : int, default=1
+        The forecasting horizon, i.e., the number of steps ahead to predict.
+
+    Attributes
+    ----------
+    data_ : list of float
```
[the remark this answers is not in the document]
Comment (alexbanwell1) [reply aeon/forecasting/_arima.py]: Done in #39a3ed2

File: aeon/utils/forecasting/_hypo_tests.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
+import numpy as np
+
+
+def kpss_test(y, regression="c", lags=None):  # Test if time series is stationary
+    """
+    Implement the KPSS test for stationarity.
```
[the remark this answers is not in the document]
Comment (alexbanwell1) [reply aeon/utils/forecasting/_hypo_tests.py]: Done in #39a3ed2

File: aeon/utils/optimisation/_nelder_mead.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    Perform optimisation using the Nelder–Mead simplex algorithm.
+
+    This function minimises a given loss (objective) function using the Nelder–Mead
+    algorithm, a derivative-free method that iteratively refines a simplex of candidate
+    solutions. The implementation supports unconstrained minimisation of functions
+    with a fixed number of parameters.
+
+    Parameters
+    ----------
+    loss_function : callable
```
[the remark this answers is not in the document]
Comment (alexbanwell1) [reply aeon/utils/optimisation/_nelder_mead.py:26]: Done in #39a3ed2

File: aeon/utils/forecasting/_seasonality.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+Includes autocorrelation function (ACF) and seasonal period estimation.
+"""
+
+import numpy as np
+from numba import njit
+
+
+@njit(cache=True, fastmath=True)
+def acf(X, max_lag):
```
[the remark this answers is not in the document]
Comment (TonyBagnall) [reply aeon/utils/forecasting/_seasonality.py]: ok, when this is in we can create an issue to do that, if we remember to do that :)

**Review by TonyBagnall — CHANGES_REQUESTED** [review]
this will need some testing, esp for _calc_arima. 

Actually think _calc_arima hould be called _arma_forecast, and might need refactoring for easier use with predict

File: aeon/forecasting/_arima.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (TonyBagnall) [reply aeon/forecasting/_arima.py]: not sure about this, we should discuss again, it implies the base class knows about the sub class contract, which is backward imo.

File: aeon/forecasting/_arima.py [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+        ----------
+        y : np.ndarray, default = None
+            A time series to predict the next horizon value for. If None,
+            predict the next horizon value after series seen in fit.
+        exog : np.ndarray, default =None
+            Optional exogenous time series data assumed to be aligned with y
+
+        Returns
+        -------
+        array[float]
```
Comment (TonyBagnall) [inline aeon/forecasting/_arima.py]: returns a float not an array

**Comment by TonyBagnall** [comment]
I have optimised the code for numba, and it is now blazingly fast at fitting.
image

and longer series

image

these results are from randomly generated stationary AR models with p 0 to 3 and q 0 to 3. For each size of n, repeat m times:
randomly generate a series, fit arma model, predict one ahead. 

m set to 30 for these. I also need to do the same for iterative forecast, but these almost look too good to be true

File: pyproject.toml [hunk 12]
```diff
@@ -169,7 +169,7 @@ testpaths = "aeon"
 doctest_optionflags = [
     "NORMALIZE_WHITESPACE",
     "ELLIPSIS",
-    "FLOAT_CMP",
+#    "FLOAT_CMP",
```
Comment (chrisholder) [inline pyproject.toml]: Is this meant to be left commented out?

Comment (TonyBagnall) [reply pyproject.toml]: good spot, thanks
